File: repository/src/main/java/by/rekuts/travelagency/repository/UserSpecification.java
package by.rekuts.travelagency.repository;

import by.rekuts.travelagency.domain.User;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.jpa.domain.Specification;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
public class UserSpecification implements Specification<User> {
    Integer id;
    String login;

    public UserSpecification(Integer id) {
        this.id = id;
    }

    @Override
    public Predicate toPredicate(Root<User> root, CriteriaQuery<?> query, CriteriaBuilder builder) {
        List<Predicate> predicates = new ArrayList<>();
        if (id != null) {
            predicates.add(
                    builder.equal(root.get("id"), id)
            );
        }
        if (login != null) {
            predicates.add(
                    builder.equal(root.get("login"), login)
            );
        }
        return builder.and(predicates.toArray(new Predicate[0]));
    }
}

File: repository/src/main/scala/by/rekuts/travelagency/repository/UserSpecification.scala
package by.rekuts.travelagency.repository

import java.util

import by.rekuts.travelagency.domain.User
import javax.persistence.criteria.{CriteriaBuilder, Predicate, Root}

class UserSpecification extends Specification[User] {
  def this(id: Int) {
    this()
    this.id = id
  }

  var login: String = _

  override def getPredicates(root: Root[User], builder: CriteriaBuilder): util.List[Predicate] = {
    var predicates = super.getPredicates(root, builder)
    if (login != null) {
      predicates.add(
        builder.equal(root.get("login"), login)
      )
    }
    predicates
  }
}
